src/replace_allocate.cpp

```diff
@@ -0,0 +1,116 @@
+#include <migraphx/replace_allocate.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/op/allocate.hpp>
+#include <map>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+std::unordered_map<instruction_ref, std::string> create_output_names(const module& mod)
+{
+    std::unordered_map<instruction_ref, std::string> prog_output_names{};
+    auto last = instruction::get_output_alias(std::prev(mod.end()));
+    if(last->name() == "@return")
+    {
+        const auto& prog_outputs = last->inputs();
+        std::vector<instruction_ref> outputs_alias(prog_outputs.size());
+
+        std::transform(prog_outputs.begin(),
+                       prog_outputs.end(),
+                       outputs_alias.begin(),
+                       [](const auto& i) { return instruction::get_output_alias(i); });
+
+        std::size_t index = 0;
+        for(auto ins : outputs_alias)
+        {
+            prog_output_names[ins] = mod.name() + ":#output_" + std::to_string(index++);
+        }
+    }
+    return prog_output_names;
+}
+
+void insert_if_allocations(instruction_ref ins, module& mod, const allocation_model& model)
+{
+    std::vector<instruction_ref> inputs = ins->inputs();
+    std::vector<module_ref> mod_args    = ins->module_inputs();
+
+    std::map<std::string, shape> name_shapes;
+    for(const auto& smod : mod_args)
+    {
+        auto ps = smod->get_parameter_shapes();
+        name_shapes.insert(ps.begin(), ps.end());
+    }
+
+    bool ins_output_allocated = false;
+    for(auto& pn : name_shapes)
+    {
+        const auto& s = pn.second;
+        instruction_ref output{};
+        if(s == ins->get_shape() and not ins_output_allocated)
```

Comment:
Do we need this condition ? Regardless of `ins_output_allocated` we always `allocate` both in `then` and `else` condition. 

We are not using `ins_output_allocated` flag later either.

Reply:
Yeah you're right, I transferred a lot of this code from lowering.cpp. Removed this condition.